pypesto/optimize/ess/cess.py

```diff
@@ -0,0 +1,340 @@
+            # meaningful? this is the overall time, and identical for all
+            #  reported points
+            'time': time.time() - self.starttime,
+            # TODO
+            # 'n_fval': self.evaluator.n_eval,
+            'optimizer': str(self),
+        }
+        i_result = 0
+        result = pypesto.Result(problem=problem)
+
+        # save global best
+        optimizer_result = pypesto.OptimizerResult(
+            id=str(i_result),
+            x=self.x_best,
+            fval=self.fx_best,
+            message="Global best",
+            **common_result_fields,
+        )
+        # TODO DW: Create a single History with the global best?
+        result.optimize_result.append(optimizer_result)
+
+        # save refsets
+        for i_refset, refset in enumerate(refsets):
+            for i in range(refset.dim):
+                i_result += 1
+                result.optimize_result.append(
+                    pypesto.OptimizerResult(
+                        id=str(i_result),
+                        x=refset.x[i],
+                        fval=refset.fx[i],
+                        message=f"RefSet[{i_refset}][{i}]",
+                        **common_result_fields,
+                    )
+                )
+
+        # TODO DW: also save local solutions?
+        #  (need to track fvals or re-evaluate)
+
+        return result
+
+    def _run_scatter_searches(
+        self,
+        problem: Problem,
+        startpoint_method: StartpointMethod,
+        refsets: List[RefSet],
+    ) -> List[ESSOptimizer]:
+        """Start all scatter searches in different processes."""
+        # set default value of max_eval if not present.
+        #  only set it on a copy, as the original dicts may be re-used
+        #  for different optimization problems.
+        # reasonable value proposed in [VillaverdeEge2012]:
+        #  2.5 < tau < 3.5, default: 2.5
+        ess_init_args = [
+            dict(
+                {'max_eval': int(10**2.5 * problem.dim)},
+                **ess_kwargs,
+            )
+            for ess_kwargs in self.ess_init_args
+        ]
+
+        ctx = multiprocessing.get_context('spawn')
+        with ctx.Pool(len(self.ess_init_args)) as pool:
+            ess_optimizers = pool.starmap(
+                self._run_single_ess,
+                (
+                    [ess_kwargs, problem, startpoint_method, refset]
+                    for (ess_kwargs, refset) in zip(ess_init_args, refsets)
+                ),
+                chunksize=1,
+            )
+        return list(ess_optimizers)
+
+    def _run_single_ess(
+        self,
+        ess_kwargs,
+        problem: Problem,
+        startpoint_method: StartpointMethod,
+        refset: RefSet,
+    ) -> ESSOptimizer:
+        """
+        Run ESS.
+
+        Helper for `starmap`.
+        """
+        # different random seeds per prcess?
+        np.random.seed((os.getpid() * int(time.time() * 1000)) % 2**32)
```

Comment:
setting the random seed (which is otherwise inherited) new per process would be good. is it possible to instead use an own random number generator, as np.random.seed is deprecated?

Reply:
I choose this because it seems to be the one used for starting point sampling in pypesto which is the most relevant RNG application here.

```
grep -r np.random pypesto/startpoint/
pypesto/startpoint/latin_hypercube.py:        xs[:, i_dim] = np.random.permutation(np.arange(n_starts))
pypesto/startpoint/latin_hypercube.py:        xs += np.random.random((n_starts, dim))
pypesto/startpoint/uniform.py:    xs = np.random.random((n_starts, dim))
```

Seeding any other RND would have no effect, unless this is also changed elsewhere. Maybe replace all of them separately later?

Reply:
Ah, you mean being able to supply any other random regenerator? I guess that would only be relevant in combination with a custom starting point method that uses a different RNG than the rest of pypesto. Could be done. But I'd generalize that then to an arbitrary init function that can be passed, independent of RNGs.

Reply:
ah if it's the pypesto starting point generation, then yes we would need to change it there too. good to leave for the moment, probably at some point we should go over all np.random samplers and supply proper random number generators.